### PR TITLE
[opentitantool] Proxy transport implementation

### PR DESCRIPTION
**Pull request by jesultra**
Add a new implementation of the Transport trait, which forwards every
method invocation via a network protocol to a separate session proxy
process.

This is to be used either with the session process being a software
emulator of sorts, or with the session process managing a USB
connection to a debugger device connected to hardware.

This is one of the steps outlined in issue https://github.com/lowRISC/opentitan/issues/10889.  Implementation of
the session process will come in a future commit. (PR #11054)

Signed-off-by: Jes B. Klinke <jbk@chromeos.org>

### CODE REVIEW

File: sw/host/opentitanlib/src/transport/proxy/spi.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            _ => Err(TransportError::ProxyCommunicationError(
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    fn run_transaction(&self, transaction: &mut [Transfer]) -> Result<()> {
+        let mut req: Vec<SpiTransferRequest> = Vec::new();
+        for transfer in &*transaction {
+            // &* to treat as non-mutable in this loop
```
Comment (mcy) [inline sw/host/opentitanlib/src/transport/proxy/spi.rs:127]: Why is this necessary? It doesn't seem obvious to me.

File: sw/host/opentitanlib/src/transport/proxy/mod.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+use crate::transport::{
+    Capabilities, Capability, Result, Transport, TransportError, WrapInTransportError,
+};
+
+mod gpio;
+mod i2c;
+mod spi;
+mod uart;
+
+/// Implementation of the Transport trait backed by connection to a remost OpenTitan tool
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/transport/proxy/mod.rs:39]: nit: s/remost/remote

File: sw/host/opentitanlib/src/transport/proxy/gpio.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    // Convenience method for issuing GPIO commands via proxy protocol. 
+    fn execute_command(&self, command: GpioRequest) -> Result<GpioResponse> {
+        match self.inner.execute_command(Request::Gpio {
+            id: self.pinname.clone(),
+            command,
+        })? {
+            Response::Gpio(resp) => Ok(resp),
+            _ => Err(TransportError::ProxyCommunicationError(
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/gpio.rs:34]: I have made ProxyCommunicationError contain an enum, just as we have GpioError, SpiError, etc.

File: sw/host/opentitanlib/src/transport/proxy/gpio.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                "Unexpected reply".to_string(),
+            )),
+        }
+    }
+
+    /// Sets the value of the GPIO pin `id` to `value`.
+    fn write(&self, value: bool) -> Result<()> {
+        match self.execute_command(GpioRequest::Write { logic: value })? {
+            GpioResponse::Write => Ok(()),
+            _ => bail!(TransportError::ProxyCommunicationError(
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/gpio.rs:56]: I like the fact that with bail!() I do not need to add .into() at the end, when directly using the new ProxyError enum which implements Into<TransportError>.

File: sw/host/opentitanlib/src/transport/proxy/mod.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        Ok(serde_json::from_slice::<Message>(&buf[..idx - 1])?)
+    }
+}
+
+impl Transport for Proxy {
+    fn capabilities(&self) -> Capabilities {
+        Capabilities::new(Capability::UART | Capability::GPIO | Capability::SPI | Capability::I2C)
+    }
+
+    // Crate SPI Target instance, or return one from a cache of previously created instances.
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/transport/proxy/mod.rs:115]: nit: s/Crate/create ? applies throughout

File: sw/host/opentitanlib/src/transport/proxy/mod.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            let rc = conn.read(&mut buf[idx..])?;
+            if rc == 0 {
+                anyhow::bail!(io::Error::new(ErrorKind::UnexpectedEof, "Truncated JSON"))
+            }
+            idx += rc;
+            if buf[idx - 1] == b'\n' {
+                break;
+            }
+        }
+        Ok(serde_json::from_slice::<Message>(&buf[..idx - 1])?)
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/mod.rs:93]: Without it, I get the error below.  I guess that serde functions do not use anyhow, but rather their specific error enum.
```
expected struct `anyhow::Error`, found struct `serde_json::Error`
```

File: sw/host/opentitanlib/src/transport/proxy/spi.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/spi.rs:127]: Without the ```&*```, I get this error:
```
105 |         let mut req: Vec<SpiTransferRequest> = Vec::new();
106 |         for transfer in transaction {
    |                         -----------
    |                         |
    |                         `transaction` moved due to this implicit call to `.into_iter()`
    |                         help: consider borrowing to avoid moving into the for loop: `&transaction`
...
129 |                     resp.len() == transaction.len(),
    |                                   ^^^^^^^^^^^^^^^^^ value borrowed here after move
```

If I add just a single ```&``` as suggested above, I instead get:
```
106 |         for transfer in &transaction {
    |                         ^^^^^^^^^^^^ `&&mut [io::spi::Transfer<'_, '_>]` is not an iterator
```

So that is why I ended up with the strangely looking ```&*``` to temporarily convert the mutable reference into an immutable reference.  (Maybe there is a more explicit way of doing the same, that I am not aware of.)

File: sw/host/opentitanlib/src/proxy/protocol.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+#[derive(Serialize, Deserialize)]
+pub enum Message {
+    Req(Request),
+    Res(Result<Response, TransportError>),
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum Request {
+    Gpio { id: String, command: GpioRequest },
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/proxy/protocol.rs]: This file was part of the "prerequisite commit" that I had included in the initial PR because GitHub has no support for a chain of dependent PRs being reviewed concurrently.  The file has now been removed from this PR, as the other PR has merged.

**Review by timothytrippel — APPROVED** [review]

File: sw/host/opentitanlib/src/transport/proxy/mod.rs (hunk 5 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/mod.rs]: Argh.  9 out of 10 times when I need to type "use crate::...", I end up typing "create" instead.  It seems that my fingers cannot even get one of the two right consistently.

File: sw/host/opentitanlib/src/transport/proxy/mod.rs (hunk 2 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/mod.rs]: Done

File: sw/host/opentitanlib/src/transport/proxy/spi.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/spi.rs:107]: This also works, may that is slightly better than `&*`:
```
for transfer in transaction.iter() {
```

Comment (mcy) [reply sw/host/opentitanlib/src/transport/proxy/spi.rs:107]: Neat error! Yeah iter() is probably nicer all around.

**Review by mcy — APPROVED** [review]

File: sw/host/opentitanlib/src/backend/mod.rs [hunk 8]
```diff
@@ -49,7 +53,8 @@ pub enum Error {
 /// Creates the requested backend interface according to [`BackendOpts`].
 pub fn create(args: &BackendOpts) -> Result<TransportWrapper> {
     let mut env = TransportWrapper::new(match args.interface.as_str() {
-        "" => create_empty_transport(),
+        "proxy" | "" => proxy::create(&args.proxy_opts),
```
Comment (timothytrippel) [inline sw/host/opentitanlib/src/backend/mod.rs:56]: One last item (as discussed offline), since this is re-used by the session process, we probably want to have a different default for opentitantool vs opentitansession, since the session should connect to a real / emulated OT device, rather than another session.

File: sw/host/opentitanlib/src/transport/proxy/spi.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/spi.rs]: Done.

File: sw/host/opentitanlib/src/backend/mod.rs (hunk 8 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/backend/mod.rs]: Done

**Comment by jesultra** [comment]
I have addressed all comments, and think this is ready to be merged, once CI completes.
